keps/sig-cloud-provider/0013-removing-in-tree-providers.md

```diff
@@ -0,0 +1,272 @@
+Kubernetes release cycle. The de-coupling of provider code allows for separation of concern between the core Kubernetes
+and the initial cloud providers that are currently built into the project. Additionally, one of the goals of SIG Cloud
+Provider is to move all providers out-of-tree to promote a vendor neutral ecosystem.
+
+### Goals
+
+- Remove all cloud provider specific code from kubernetes/kubernetes.
+- Move the cloud provider specific controller manager logic into repos appropriate for those cloud providers.
+
+### Non-Goals
+
+- Testing/validating out-of-tree provider code for all cloud providers, this should be done by providers
+
+## Proposal
+
+### Approach
+
+In order to remove cloud provider code from kubernetes/kubernetes. A 3 phase approach will be taken.
+
+1. Move all code in `kubernetes/kubernetes/pkg/cloudprovider/providers` to `kubernetes/kubernetes/staging/cloud-provider-<provider>/`. This requires removing all dependencies in each cloud provider to `kubernetes/kubernetes`.
+2. Begin to build CCM from external repos (`kubernetes/cloud-provider-<provider>`) via code synced from `kubernetes/kubernetes/staging/cloud-provider-<provider>`. This allows us to develop providers "out-of-tree" while being able to build both in-tree components like kube-controller-manager and out-of-tree components like cloud-controller-manager. Development is still done in `kubernetes/kubernetes/staging/cloud-provider-<provider>`.
+3. Delete all code in `kubernetes/kubernetes/staging/cloud-provider-<provider>` and shift main development to `kubernetes/cloud-provider-<provider>`.
+
+#### Phase 1 - Moving Cloud Provider Code to Staging
+
+In Phase 1, all cloud provider code in `kubernetes/kubernetes/pkg/cloudprovider/providers` will be moved to `kubernetes/kubernetes/staging/cloudprovider/providers`. Moving code to a staging directory still allows us to build core Kubernetes components (`kube-controller-manager`, `kubelet`, etc) without interruption while also signalling to the community that in-tree provider code is slated for removal.
+
+The biggest challenge of this phase is to remove dependences to kubernetes/kubernetes in all the providers. This is required to avoid circular dependencies within the various Kubernetes repos. All other repos "staged" (`client-go`, `apimachinery`, `api`, etc) in Kubernetes follow the same pattern. Below are the current list of packges in `kubernetes/kubernetes` that we need to removed. Note that _some_ dependencies may need to be duplicated for the duration of this migration as other critical components of Kubernetes may share the same code but removing the dependency may not be trivial. In general, we will avoid duplicating code and only use it as a last resort to resolve circular dependencies to `kubernetes/kubernetes`.
+
+| Dependency                        | Proposed Location            | Is duplicated? | Partial or Full? |
```

Comment:
Need help filling this out, if you are assigned to any of the dependencies in https://github.com/kubernetes/kubernetes/issues/69585 please comment here on what the new locations should be :) thanks!

cc @kubernetes/sig-cloud-provider @cheftako @dims @d-nishi @frapposelli @hogepodge @islinwb @mcrute @stewart-yu @yastij

Reply:
Realizing now that the `Partial vs Full` part is confusing, but this just indicating whether we need to migrate the entire package for our purposes or if we can just pick out the specific methods/types that we need.

Reply:
`pkg/controller` need more investgate, some api need move from `k8s.io/kubernetes/pkg/apis/core`, and code from "k8s.io/kubernetes/pkg/apis/core/helper" . i'll update later

Reply:
For `pkg/controller`, it seems like we already copied [ControllerClientBuilder](https://github.com/kubernetes/cloud-provider/blob/master/cloud.go#L31-L40) and the method `AddOrUpdateTaintOnNode` can be moved to `k8s.io/cloud-provider/nodes`. What do you think?

Reply:
ah, sorry I forgot about dependency to `k8s.io/kubernetes/pkg/util/taints`

Reply:
Seems like `AddOrUpdateTaint` is only used in `pkg/controller` so might be reasonable to just move that method into `k8s.io/cloud-provider/nodes` as well.

Reply:
i move part of  `taint` op from  `k8s.io/kubernetes/pkg/util/taints` to `k8s.io/apimachinery/pkg/util/taints`. see 
https://github.com/kubernetes/kubernetes/pull/70225

Reply:
so form table as below:
`pkg/controller`          `k8s.io/cloud-provider/pkg/controller`         `Yes`          `Partial`

Reply:
Is the `pkg` necessary? Can it just be `k8s.io/cloud-provider/controller`?

Reply:
`k8s.io/cloud-provider/controller` ok for me